### PR TITLE
Adding Leo's example

### PR DESCRIPTION
**Pull request by mcoughlin**
Leo's example of intersecting points with localizations: https://github.com/growth-astro/healpix-intersection-example is needed for SkyPortal. This makes it possible to add the healpix index to each point.

### CODE REVIEW

**Comment by mcoughlin** [comment]
@lpsinger Of course. I just dont have the power to request one from you!

**Comment by lpsinger** [comment]
> @lpsinger Of course. I just dont have the power to request one from you!

Uh oh. Why not? Do I need to add you to the project?

**Comment by mcoughlin** [comment]
@lpsinger Yup!

**Comment by lpsinger** [comment]
This looks like a faithful migration of the code that @dannygoldstein and I worked on back to this repository, but my main concern about merging it is that we never reached closure on the strategy for representing raster data sets (i.e., sky maps). I would like to see the schema for that too.

File: healpix_alchemy/tests/test_regions.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class Field(Region, Base):
+    __tablename__ = 'fields'
+
+    field_id = Column(
+        Integer,
+        primary_key=True)
+
+    tile_class = lambda: FieldTile  # pylint: disable=E731 # noqa
```
Comment (lpsinger) [inline healpix_alchemy/tests/test_regions.py:26]: Why both the pylint and noqa directives?

File: healpix_alchemy/healpix.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            self.uniq = uniq
+
+    @hybrid_property
+    def uniq(self):
+        """HEALPix UNIQ pixel index."""
+        # This is the same expression as in astropy_healpix.level_ipix_to_uniq,
+        # but reproduced here so that SQLAlchemy can map it to SQL.
+        ipix = self.nested_lo
+        # FIXME: wrong, get level from log2(nested_hi - nested_lo)/2
+        return int(ipix + (1 << 2 * (LEVEL + 1)))
```
Comment (lpsinger) [inline healpix_alchemy/healpix.py:49]: Indeed, this is wrong.

**Comment by mcoughlin** [comment]
@lpsinger the current cross match PR is here: https://github.com/skyportal/skyportal/pull/2129.

It seems to work pretty well.

File: healpix_alchemy/tests/test_regions.py (hunk 1 above)
Comment (mcoughlin) [reply healpix_alchemy/tests/test_regions.py]: unnecessary! removed.

File: healpix_alchemy/healpix.py (hunk 2 above)
Comment (mcoughlin) [reply healpix_alchemy/healpix.py]: Ok updated with the expression in that line.

File: healpix_alchemy/tests/test_regions.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class Field(Region, Base):
+    __tablename__ = 'fields'
+
+    field_id = Column(
+        Integer,
+        primary_key=True)
+
+    tile_class = lambda: FieldTile # noqa
```
Comment (lpsinger) [inline healpix_alchemy/tests/test_regions.py:26]:
```suggestion
    tile_class = lambda: FieldTile  # noqa: E731
```

Always specify the warning that is being overridden. (For the same reason that you should always specify the exception that is to be caught).

Also, PEP8 says:

> An inline comment is a comment on the same line as a statement. Inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space.

File: healpix_alchemy/healpix.py (hunk 2 above)
Comment (lpsinger) [reply healpix_alchemy/healpix.py]: Still wrong... this script:

```python
#!/usr/bin/env python
from astropy_healpix import uniq_to_level_ipix, level_ipix_to_uniq
from mocpy import MOC
import numpy as np

LEVEL = MOC.HPY_MAX_NORDER

value = level_ipix_to_uniq(3, 12)
print('expected', value)

# @uniq.setter
level, ipix = uniq_to_level_ipix(value)
shift = 2 * (LEVEL - level)
nested_lo = int(ipix << shift)
nested_hi = int(((ipix + 1) << shift) - 1)

# @uniq
ipix = nested_lo
level = int(np.log2(nested_hi - nested_lo)/2)
uniq = int(ipix + (1 << 2 * (level + 1)))

print('got', uniq)
```

produces this output:

```
expected 268
got 72057594037927936
```

 The expression has to involve both `level` and `LEVEL`, like in the setter.

**Comment by lpsinger** [comment]
> @lpsinger the current cross match PR is here: skyportal/skyportal#2129.
>
> It seems to work pretty well.

That is doing the cross matching in Python. I think that in order to be able to do real-time, interactive queries with a cross-match between a sky map and all of the ZTF tiles, or between a sky map and a galaxy catalog, it needs to be done inside the database. It's not computationally demanding, it's just too much data to shift between the database and the application.

The trick will be to recognize that both the sky map and the multi-resolution HEALPix tiles are represented as collections of intervals. With the appropriate representation, the database should be able to combine any number of multi-resolution HEALPix data sets very quickly because it amounts to merging N sorted lists.

**Comment by mcoughlin** [comment]
Can you explain more here Leo? Isn't the fact it is a sub query mean it is in the database? How should it look to ensure it is?

**Comment by lpsinger** [comment]
You are precomputing the probability contained in each field and then loading that into the database. But I think that we need to be able to calculate the integrated probability within an arbitrary set of fields, correctly accounting for overlap, within the database, so that the user can interactively edit the sky map and see the probability update.

**Comment by mcoughlin** [comment]
The cumulative probability for a given skymap will not change right? I of course understand the need for speed, but a precomputed set of calculations upon skymap ingestion seems like a reasonable tradeoff for difficulty of technical development right? Unless you think the same calculation in sql is trivial?

**Comment by lpsinger** [comment]
Correct. But that's not what I'm concerned about. I want the following operations to be made fast using the databases indices:

1. [x] Match raster with points (evaluate sky map at positions in a galaxy catalog)
2. [x] Match regions with points (find galaxy catalog entries within a set of fields)
3. [ ] Match raster with regions and reduce (find total probability contained within a set of fields)
4. [ ] Match regions with regions and reduce (find total area within a set of fields)

We are currently lacking queries or algorithms for the last two.

**Comment by mcoughlin** [comment]
But that's a completely separate PR right? We are just dealing with object crossmatch here.

**Comment by lpsinger** [comment]
No, it's not. I am far from certain that this schema is the correct one to facilitate the last two operations. That is why I was working on this demo in a separate sandbox repository.

**Comment by mcoughlin** [comment]
Ok and what is the main issue with the field coverage calculation? Playing around with it here seemed ok https://github.com/skyportal/skyportal/pull/2133

File: healpix_alchemy/healpix.py (hunk 2 above)
Comment (mcoughlin) [reply healpix_alchemy/healpix.py]: How about something like the below?

`LEVEL = MOC.HPY_MAX_NORDER

value = level_ipix_to_uniq(3, 12)
print('expected', value)

# @uniq.setter
level, ipix = uniq_to_level_ipix(value)
shift = 2 * (LEVEL - level)
nested_lo = int(ipix << shift)
nested_hi = int(((ipix + 1) << shift) - 1)

# @uniq
level = LEVEL - (int(np.ceil(np.log2(nested_hi - nested_lo)/2)))
shift = 2 * (LEVEL - level)
ipix = int(nested_lo >> shift)
uniq = int(ipix + (1 << 2 * (level + 1)))
print('got', uniq)`

Comment (lpsinger) [reply healpix_alchemy/healpix.py]: Looks plausible. How about a unit test?

**Comment by mcoughlin** [comment]
Ah its in a PR based on that PR ;) 

https://github.com/skyportal/skyportal/pull/2157/files#diff-a34c5cc426ca4c36e2ded58763848664ee2125c42dacfe8ee83b84cadc72cdd2R267

File: healpix_alchemy/tests/test_regions.py [hunk 4]
```diff
@@ -101,3 +102,23 @@ def test_fields(request, session, engine):
...
+
+def test_tile(engine):
+    """Generate an example Tile and confirm index calculation."""
+    Base.metadata.create_all(engine)
+
+    LEVEL = MOC.HPY_MAX_NORDER
+
+    value = level_ipix_to_uniq(3, 12)
+
+    # @uniq.setter
```
Comment (lpsinger) [inline healpix_alchemy/tests/test_regions.py:115]: The unit test should check the methods themselves, not a copies of them.

Comment (mcoughlin) [reply healpix_alchemy/tests/test_regions.py:115]: But I am checking the Tile output right based on the calculation? Maybe I don't understand what you mean? Or do you just want the values of nested_lo and nested_hi hardcoded?

**Comment by lpsinger** [comment]
> Ah its in a PR based on that PR ;)
> 
> https://github.com/skyportal/skyportal/pull/2157/files#diff-a34c5cc426ca4c36e2ded58763848664ee2125c42dacfe8ee83b84cadc72cdd2R267

Right... [this part](https://github.com/skyportal/skyportal/pull/2157/files#diff-a34c5cc426ca4c36e2ded58763848664ee2125c42dacfe8ee83b84cadc72cdd2R267-R294) gives you the probabilities contained within each field in a list, but I think that we still need a query to calculate the probability within the union of those fields.

File: healpix_alchemy/tests/test_regions.py (hunk 4 above)
Comment (lpsinger) [reply healpix_alchemy/tests/test_regions.py:115]: This unit test does show that your method is correct, but I would like you to unit test the actual getter and setter methods, not simply copy-paste their bodies into the unit test.

**Comment by mcoughlin** [comment]
Presumably that isn't required until there is a frontend element that can display it, but I guess your concern is that such a query cannot be composed easily in the current schema?

**Comment by mcoughlin** [comment]
Ok this makes sense; we are in agreement here.

**Comment by mcoughlin** [comment]
So right now, we have something like:

    level, ipix = 3, 12
    value = level_ipix_to_uniq(level, ipix)
    shift = 2 * (LEVEL - level)
    nested_lo = int(ipix << shift)
    nested_hi = int(((ipix + 1) << shift) - 1)
    
To calculate the nested_lo and nested_hi to compare against: do you mean some other method to check how these are generated?

**Comment by lpsinger** [comment]
> So right now, we have something like:
> 
> ```
> level, ipix = 3, 12
> value = level_ipix_to_uniq(level, ipix)
> shift = 2 * (LEVEL - level)
> nested_lo = int(ipix << shift)
> nested_hi = int(((ipix + 1) << shift) - 1)
> ```
> 
> To calculate the nested_lo and nested_hi to compare against: do you mean some other method to check how these are generated?

I am looking for a test that the uniq getter and setter methods round trip.

**Comment by mcoughlin** [comment]
This tests the setter:
    assert tile.nested_lo == nested_lo
    assert tile.nested_hi == nested_hi
and this sets the getter:
    assert tile.uniq == value
right? I am just clearly missing something

**Comment by lpsinger** [comment]
> This tests the setter:
> assert tile.nested_lo == nested_lo
> assert tile.nested_hi == nested_hi
> and this sets the getter:
> assert tile.uniq == value
> right? I am just clearly missing something

Oh! You're right. Sorry for being thick.

File: healpix_alchemy/healpix.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        super().__init__(*args, **kwargs)
+        if uniq is not None:
+            self.uniq = uniq
+
+    @hybrid_property
+    def uniq(self):
+        """HEALPix UNIQ pixel index."""
+        # This is the same expression as in astropy_healpix.level_ipix_to_uniq,
+        # but reproduced here so that SQLAlchemy can map it to SQL.
+        level = LEVEL - (int(np.ceil(np.log2(self.nested_hi -
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply healpix_alchemy/healpix.py:48]: So something like this is plausible:
        level = LEVEL - (func.ceil((func.log(self.nested_hi -
                                             self.nested_lo)/2)/func.log(2)))
But the bitshifting is likely to be an issue?
http://slavasql.blogspot.com/2019/12/t-sql-bitwise-shifting.html

**Comment by mcoughlin** [comment]
@lpsinger Instead of the uniq getter, why not just store uniq as a column?

File: healpix_alchemy/healpix.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lpsinger) [reply healpix_alchemy/healpix.py:48]: Postgres supports bit string operations: https://www.postgresql.org/docs/current/functions-bitstring.html

The question is, does SQLAlchemy know about it?

**Comment by lpsinger** [comment]
> @lpsinger Instead of the uniq getter, why not just store uniq as a column?

It's redundant information; it would increase the size of the database.

Maybe it doesn't need to be an ORM property at all --- maybe it is fine if it's pure Python.

**Comment by mcoughlin** [comment]
So we can consider leaving it as is then? Obviously is easier.

**Comment by lpsinger** [comment]
> So we can consider leaving it as is then? Obviously is easier.

I _think_ so. But we do need to change it from a `@hybrid_property` to a `@property`. That should be alright because it's only ever used when initializing a row.

**Comment by mcoughlin** [comment]
I pushed this change, sorry for the delay.

**Comment by mcoughlin** [comment]
@lpsinger what else do we need to get this in?

**Comment by lpsinger** [comment]
The `Point` class needs to have the `nested` column correctly initialized from the right ascension and declination.

It didn't have that initialization in my original example code, because the example code only stored HEALPix in the database, not right ascension and declination.

**Comment by mcoughlin** [comment]
Ok is that what you had in mind?

File: healpix_alchemy/point.py [hunk 6]
```diff
@@ -20,7 +22,11 @@ def __init__(self, *args, ra=None, dec=None, nested=None, **kwargs):
         super().__init__(*args, **kwargs)
         self.ra = ra
         self.dec = dec
-        self.nested = nested
+        if nested is not None:
+            self.nested = nested
+        else:
+            self.nested = HPX.lonlat_to_healpix(self.ra * u.deg,
+                                                self.dec * u.deg)
```
Comment (lpsinger) [inline healpix_alchemy/point.py:29]:
```suggestion
        if nested is None:
            nested = HPX.lonlat_to_healpix(ra * u.deg, dec * u.deg)
        self.nested = nested
```

**Comment by mcoughlin** [comment]
Thanks for that Leo. Anything else?

**Review by lpsinger — APPROVED** [review]

**Comment by mcoughlin** [comment]
Thanks Leo! Are you comfortable doing a pypi release so we can use that? Or should we pin to master for now?

**Comment by lpsinger** [comment]
There are a couple of old PRs I would like to get merged before a release. I'll probably be asking you to review them.

**Comment by mcoughlin** [comment]
Ok thanks for letting me know Leo!

**Pull request merged**